exec/java-exec/src/main/java/org/apache/drill/exec/rpc/AbstractUserClientConnectionWrapper.java

```diff
@@ -0,0 +1,101 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.drill.exec.rpc;
+
+import com.google.common.base.Preconditions;
+import org.apache.drill.common.exceptions.UserException;
+import org.apache.drill.common.exceptions.UserRemoteException;
+import org.apache.drill.exec.proto.GeneralRPCProtos;
+import org.apache.drill.exec.proto.UserBitShared.DrillPBError;
+import org.apache.drill.exec.proto.UserBitShared.QueryId;
+import org.apache.drill.exec.proto.UserBitShared.QueryResult;
+import org.apache.drill.exec.proto.helper.QueryIdHelper;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+public abstract class AbstractUserClientConnectionWrapper implements UserClientConnection {
+  private static final org.slf4j.Logger logger =
+      org.slf4j.LoggerFactory.getLogger(AbstractUserClientConnectionWrapper.class);
+
+  protected final CountDownLatch latch = new CountDownLatch(1);
+
+  protected volatile DrillPBError error;
+
+  protected volatile UserException exception;
+
+  /**
+   * Wait until the query has completed or timeout is passed.
+   *
+   * @throws InterruptedException
+   */
+  public boolean await(final long timeoutMillis) throws InterruptedException {
+    return latch.await(timeoutMillis, TimeUnit.MILLISECONDS);
+  }
+
+  /**
+   * Wait indefinitely until the query is completed. Used only in case of WebUser
+   *
+   * @throws Exception
+   */
+  public void await() throws Exception {
+    latch.await();
+    if (exception != null) {
+      throw exception;
+    }
+  }
+
+  @Override
+  public void sendResult(RpcOutcomeListener<GeneralRPCProtos.Ack> listener, QueryResult result) {
+
+    Preconditions.checkState(result.hasQueryState());
+
+    // Release the wait latch if the query is terminated.
+    final QueryResult.QueryState state = result.getQueryState();
+    final QueryId queryId = result.getQueryId();
+
+    if (logger.isDebugEnabled()) {
+      logger.debug("Result arrived for QueryId: {} with QueryState: {}", QueryIdHelper.getQueryId(queryId), state);
+    }
+
+    switch (state) {
+      case FAILED:
+        error = result.getError(0);
+        exception = new UserRemoteException(error);
+        latch.countDown();
+        break;
+      case CANCELED:
+      case COMPLETED:
+        Preconditions.checkState(result.getErrorCount() == 0);
+        latch.countDown();
+        break;
+      default:
+        logger.error("Query with QueryId: {} is in unexpected state: {}", queryId, state);
```

Comment:
As discussed I have opened a new [DRILL-5558](https://issues.apache.org/jira/browse/DRILL-5558) for this to be addressed in a separated PR since it's existing behavior in DrillClient as well.